### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by UjjwalBudha**
sent for review

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
+name: Deploying an app
```
Comment (mukes137) [inline .github/workflows/main.yml:1]: relevant name.
specify which app?

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+on:
+  workflow_dispatch:
+  push:
+    branches: ["dev", "main"]
```
Comment (mukes137) [inline .github/workflows/main.yml:6]: why are we deploying app both on push and on pull request for main?

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  pull_request:
+    branches: ["main"]
+
+jobs:
+  dev:
+    runs-on: ubuntu-latest
+    env:
```
Comment (mukes137) [inline .github/workflows/main.yml:13]: specify env variables globally as mush as possible.

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+          python-version: '3.x'
+
+      - name: Install Ansible and AWS collections
+        run: |
+          python -m pip install --upgrade pip
+          pip install ansible boto3 botocore
+          ansible-galaxy collection install amazon.aws
+
+      - name: Configure AWS Profile
+        run: |
```
Comment (mukes137) [inline .github/workflows/main.yml:51]: why are we setting profile?

File: ansible/playbook.yml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+---
+- name: configure frontend servers
+  hosts: _frontend_dev_ujwal_final_project
```
Comment (mukes137) [inline ansible/playbook.yml:3]: where is this hostname coming from?
any idea?

File: ansible/roles/backend/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+- name: Restart SSM agent
+  service:
+    name: snap.amazon-ssm-agent.amazon-ssm-agent.service
+    state: restarted
+
+- name: Install NVM
+  shell: |
+    curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh | bash
+    source ~/.bashrc
```
Comment (mukes137) [inline ansible/roles/backend/tasks/main.yml:23]: is this working?
confirm it.

File: ansible/roles/backend/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+- name: Install npm packages for the backend
+  npm:
+    path: /home/ubuntu/final-project-app-backend
+    state: present
+
+- name: Deploy Nginx configuration
+  template:
+    src: nginx.conf.j2
+    dest: /etc/nginx/sites-available/default
```
Comment (mukes137) [inline ansible/roles/backend/tasks/main.yml:40]: please create other file for configuration than overwriting the default one.

File: ansible/roles/backend/templates/nginx.conf.j2 [hunk 8]
```diff
@@ -0,0 +1,27 @@
+server {
+    listen 80;
+    server_name ujwal.sandbox.adex.ltd;
+
+    location / {
+        proxy_pass http://localhost:8082;
```
Comment (mukes137) [inline ansible/roles/backend/templates/nginx.conf.j2:6]: better to use port as variables.

File: ansible/roles/frontend/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  service:
+    name: snap.amazon-ssm-agent.amazon-ssm-agent.service
+    state: restarted
+
+- name: Ensure npm is installed
+  apt:
+    name: npm
+    state: present
+
+- name: Install MySQL server
```
Comment (mukes137) [inline ansible/roles/frontend/tasks/main.yml:17]: why are we installing mysql server?
Aren't we using rds as database.

File: terraform/backend-autoscaling.tf [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  source  = "terraform-aws-modules/autoscaling/aws"
+  version = "6.1.0"
+
+  name                 = "backend-${var.environment}-${var.project_name}"
+  launch_template_name = "backend-launch-template-${var.environment}-${var.project_name}"
+
+  # Launch template
+  image_id             = var.ami
+  instance_type        = "t2.micro"
+  key_name             = "intern-ujwal"
```
Comment (mukes137) [inline terraform/backend-autoscaling.tf:11]: why are we using ssh key and instance profile both?
Use the safer option.

File: terraform/ec2.tf [hunk 11]
```diff
@@ -0,0 +1,52 @@
+# module "ec2_instance" {
```
Comment (mukes137) [inline terraform/ec2.tf:1]: remove the commented lines.

File: terraform/iam.tf [hunk 12]
```diff
@@ -0,0 +1,26 @@
+resource "aws_iam_role" "ssm_role" {
+  name = "ssm-role"
```
Comment (mukes137) [inline terraform/iam.tf:2]: follow consistent naming for all of the resources.

File: terraform/loadbalancer-backend.tf [hunk 13]
```diff
@@ -0,0 +1,87 @@
+resource "aws_lb" "backend_alb" {
```
Comment (mukes137) [inline terraform/loadbalancer-backend.tf:1]: why are we using resource for loadbalancer?

File: terraform/rds.tf [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  identifier = "database-${var.environment}-${var.project_name}"
+
+  engine            = "mysql"
+  engine_version    = "8.0.35"
+  instance_class    = "db.t3.micro"
+  allocated_storage = 5
+
+  db_name  = "products"
+  username = "ujwal"
+  password = "ujwal123"
```
Comment (mukes137) [inline terraform/rds.tf:14]: use relevant aws resource to store these secrets.

File: terraform/security_group.tf [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+module "frontend_sg" {
+  source = "terraform-aws-modules/security-group/aws"
+
+  name        = "frontend_sg"
+  description = "Security group for web server with SSH port open"
+  vpc_id      = module.vpc.vpc_id
+
+  ingress_with_cidr_blocks = [
+    {
+      from_port   = 22
```
Comment (mukes137) [inline terraform/security_group.tf:30]: is this relevant?

File: ansible/playbook.yml (hunk 5 above)
Comment (UjjwalBudha) [reply ansible/playbook.yml:3]: i have created two groups one for backend servers and one for frontend server

File: .github/workflows/main.yml (hunk 2 above)
Comment (UjjwalBudha) [reply .github/workflows/main.yml:6]: changed it to updates rather than main

File: .github/workflows/main.yml (hunk 4 above)
Comment (UjjwalBudha) [reply .github/workflows/main.yml:51]: removed it

File: terraform/ec2.tf (hunk 11 above)
Comment (UjjwalBudha) [reply terraform/ec2.tf:1]: uncommented those lines

File: terraform/iam.tf (hunk 12 above)
Comment (UjjwalBudha) [reply terraform/iam.tf:2]: made the changes

File: terraform/loadbalancer-backend.tf (hunk 13 above)
Comment (UjjwalBudha) [reply terraform/loadbalancer-backend.tf:1]: had to manage multiple rules so modules seemed confusing

File: terraform/rds.tf (hunk 14 above)
Comment (UjjwalBudha) [reply terraform/rds.tf:14]: used secrets manager

File: .github/workflows/main.yml (hunk 1 above)
Comment (UjjwalBudha) [reply .github/workflows/main.yml:1]: changed it

File: terraform/security_group.tf (hunk 15 above)
Comment (UjjwalBudha) [reply terraform/security_group.tf:30]: used relevant descriptions

**Pull request merged**